datafusion/physical-plan/src/topk/mod.rs

```diff
@@ -319,19 +343,89 @@ impl TopK {
     /// (a > 2 OR (a = 2 AND b < 3))
     /// ```
     fn update_filter(&mut self) -> Result<()> {
-        let Some(filter) = &self.filter else {
+        // If the heap doesn't have k elements yet, we can't create thresholds
+        let Some(max_row) = self.heap.max() else {
             return Ok(());
         };
-        let Some(thresholds) = self.heap.get_threshold_values(&self.expr)? else {
+
+        let new_threshold_row = &max_row.row;
+
+        // Extract scalar values BEFORE acquiring lock to reduce critical section
+        let thresholds = match self.heap.get_threshold_values(&self.expr)? {
```

Comment:
I think we can move this down after the check for update too:
- https://github.com/pydantic/datafusion/pull/37